### DISCUSSION TITLE
Strategy for storing snapshots

### DISCUSSION THREAD

**Discussion by buscarini** [pointfreeco/swift-snapshot-testing#504] [Q&A] [answered]
I'm wondering what would be a good strategy for storing snapshots. I started using this and I ended with a repository that was more than 2 GB, which is the maximum supported by BitBucket, so I had to create a different repo and disable all snapshot testing. ¿Do you store this only in a CI machine? ¿Can you specify a different root folder so you can have snapshots saved outside of the current repository?

**Comment by stephencelis** [comment, marked as answer]
@buscarini Have you looked into git-lfs? https://git-lfs.github.com

**Comment by Sherlouk** [comment]
We're actually going through a very similar conversation internally at the moment - we've got solutions varying from LFS (as Stephen mentioned above) and Git Submodules down to S3/cloud storage and internal CI caches.

@stephencelis Do you have any more details on how exactly you'd setup LFS for this?

**Comment by stephencelis** [comment]
@Sherlouk I think you could probably do something like `git lfs track "*/__Snapshots__/*.png"`. We've been meaning to adopt it on some of our own repos but haven't gotten around to it yet.

**Comment by tikekar** [comment]
I am trying to use git lfs for storing images. But after that the snapshot tests are failing on our CI/CD pipeline.

**Comment by tdrhq** [comment]
May I recommend https://screenshotbot.io for this purpose? Also, open-source at https://github.com/screenshotbot/screenshotbot-oss if that's your style. Not only will it store your images, but you'll not have to manually record snapshots, you'll just get notified on your Pull Requests when screenshots change.

iOS specific documentation for integrating with swift-snapshot-testing: https://screenshotbot.io/documentation/platforms/ios-apps

**Comment by mgrider** [comment]
Late to this conversation, but for my current project, we have a separate repo for snapshot test results. This means I had to create a custom snapshot function, but that wasn't such a hardship, and I've been very happy with this solution, as it means all the commits in the snapshot repo are changes to snapshots, so the designer can browse them and see nice diffs of all the screens on github. If the repo ever gets too large, I can probably either just move to a new repo, or delete the history and start over.

Another benefit is that my custom function checks for the presence of the snapshot test folder, and if it isn't there, it just bails out without running those tests. For our CI purposes (Xcode Cloud), it didn't feel like it makes a lot of sense to run the snapshot tests every time (as they take way longer than the regular tests, and eventually we'll be paying for that time).

**Comment by doozMen** [comment]
So there is still no guide on how to store them properly without blowing up the git repo. I personally like the idea of using a separate repo. The git-lfs strategy seams to be complicated and difficult git right for pipelines. Any guide exist on this?

**Reply by mgrider** [reply]
I posted a gist here: https://gist.github.com/mgrider/796ec39e28f417fbc4784d56ab9da081

This is the parent test class I'm using in my current project. (With some names scrubbed – I replaced them with "Custom".) You'll need to add a function if you're using `SwiftUI` and not `UIViewController`s.

**Reply by tdrhq** [reply]
The easiest way is using an external tool designed for this:

a) Add your snapshots to `.gitignore`
b) On CI generate the screenshots fresh, and call a command to upload the screenshots to the external tool
c) The external tool does the diffing and notifies you on your GitHub PR

I'm particularly talking about [screenshotbot](https://screenshotbot.io) which is open-source, but there's also other open-source tools such as Workday's [Vizzy](https://github.com/workday/vizzy).

**Comment by YektaDev** [comment]
This is an old thread, but in case it’s still relevant, you might find **DVC** helpful; perhaps even as a Git LFS replacement:
- https://github.com/iterative/dvc
